### PR TITLE
ERC 20 connector

### PR DESCRIPTION
**Pull request by isra67**

### CODE REVIEW

File: erc20-connector/example/src/app.controller.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { Controller } from '@nestjs/common';
+import { AppService } from './app.service';
+import { NftController } from '../../module';
+
+@Controller()
+export class AppController extends NftController {
```
Comment (tatumio) [inline erc20-connector/example/src/app.controller.ts:6]: should not be NftController

File: erc20-connector/example/src/app.service.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import { Injectable } from '@nestjs/common';
+import { PinoLogger } from 'pino-logger';
+import { InjectPinoLogger } from 'nestjs-pino';
+import { NftService } from '../../module';
+import { celoBroadcast, Currency } from '@tatumio/tatum';
+
+@Injectable()
+export class AppService extends NftService {
```
Comment (tatumio) [inline erc20-connector/example/src/app.service.ts:8]: should not be NftService

File: erc20-connector/module/package.json [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    "access": "public"
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/tatumio/tatum-blockchain-connector/erc20-connector"
+  },
+  "bugs": "https://github.com/tatumio/tatum-blockchain-connector/",
+  "dependencies": {
+    "@nestjs/common": "^7.6.1",
+    "@tatumio/tatum": "file:../../../tatum-js",
```
Comment (tatumio) [inline erc20-connector/module/package.json:38]: move this to peerDependencies and install locally to devDependencies. Using this, tatum-js won't be included inside the target package

File: erc20-connector/module/src/Erc20Service.ts [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+      }
+      return nodes[0];
+    }
+
+    private async getClient(chain: Currency, testnet: boolean) {
+      if ([Currency.ETH, Currency.BSC, Currency.CELO].includes(chain)) {
+        return new Web3((await this.getFirstNodeUrl(chain, testnet)));
+      }
+      
+      return null
```
Comment (tatumio) [inline erc20-connector/module/src/Erc20Service.ts:70]: throw otherwise

File: erc20-connector/module/src/Erc20Service.ts [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+                txData = await prepareCeloTransferErc20SignedTransaction(testnet, body as TransferCeloOrCeloErc20Token, (await this.getFirstNodeUrl(chain, testnet)));
+                break;
+            // case Currency.TRON:
+            //     txData = await prepareTronTrc20SignedTransaction(testnet, body as TransferTronTrc20);
+            //     break;
+            default:
+                throw new Erc20Error(`Unsupported chain ${chain}.`, 'unsuported.chain');
+        }
+        if (body.signatureId) {
+            return {signatureId: await this.storeKMSTransaction(txData, chain, [body.signatureId], body['index'] || 0)};
```
Comment (tatumio) [inline erc20-connector/module/src/Erc20Service.ts:136]: body.index or undefined, not default 0. must be undefined if not present

File: erc20-connector/module/src/Erc20Service.ts [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+                txData = await prepareCeloMintErc20SignedTransaction(testnet, body as MintCeloErc20, (await this.getFirstNodeUrl(chain, testnet)));
+                break;
+            // case Currency.TRON:
+            //     // TODO!!!
+            //     break
+            default:
+                throw new Erc20Error(`Unsupported chain ${chain}.`, 'unsupported.chain');
+        }
+        if (body.signatureId) {
+            return {signatureId: await this.storeKMSTransaction(txData, chain, [body.signatureId], body['index'] || 0)};
```
Comment (tatumio) [inline erc20-connector/module/src/Erc20Service.ts:188]: not default 0

File: erc20-connector/module/src/Erc20Service.ts [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+            //         txData = await prepareTronCreateTrc20SignedKMSTransaction(testnet, body as CreateTronTrc20);
+            //     } else {
+            //         txData = await prepareTronCreateTrc20SignedTransaction(testnet, body as CreateTronTrc20);
+            //     }
+            //     break;
+            default:
+                throw new Erc20Error(`Unsupported chain ${chain}.`, 'unsuported.chain');
+        }
+        if (body.signatureId) {
+            return {signatureId: await this.storeKMSTransaction(txData, chain, [body.signatureId], body['index'] || 0)};
```
Comment (tatumio) [inline erc20-connector/module/src/Erc20Service.ts:218]: not default 0

**Review by tatumio — CHANGES_REQUESTED** [review]

File: erc20-connector/example/src/app.controller.ts (hunk 1 above)
Comment (isra67) [reply erc20-connector/example/src/app.controller.ts:6]: ok

File: erc20-connector/example/src/app.service.ts (hunk 2 above)
Comment (isra67) [reply erc20-connector/example/src/app.service.ts:8]: ok

File: erc20-connector/module/src/Erc20Service.ts (hunk 4 above)
Comment (isra67) [reply erc20-connector/module/src/Erc20Service.ts:70]: ok

File: erc20-connector/module/src/Erc20Service.ts (hunk 5 above)
Comment (isra67) [reply erc20-connector/module/src/Erc20Service.ts:136]: ok

File: erc20-connector/module/src/Erc20Service.ts (hunk 6 above)
Comment (isra67) [reply erc20-connector/module/src/Erc20Service.ts:188]: ok

File: erc20-connector/module/src/Erc20Service.ts (hunk 7 above)
Comment (isra67) [reply erc20-connector/module/src/Erc20Service.ts:218]: ok

File: erc20-connector/module/package.json (hunk 3 above)
Comment (isra67) [reply erc20-connector/module/package.json:38]: ok

File: erc20-connector/module/src/Erc20Service.ts [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+
+    private async getClient(chain: Currency, testnet: boolean) {
+      if ([Currency.ETH, Currency.BSC, Currency.CELO].includes(chain)) {
+        return new Web3((await this.getFirstNodeUrl(chain, testnet)));
+      }
+      
+      throw new Erc20Error(`Unsupported chain ${chain}.`, 'unsuported.chain');
+    }
+
+    protected broadcast(chain: Currency, txData: string, signatureId?: string) {
```
Comment (tatumio) [inline erc20-connector/module/src/Erc20Service.ts:81]: broadcast should be abstract and implemented inside core.

File: erc20-connector/module/src/Erc20Service.ts [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+                txData = await prepareCeloTransferErc20SignedTransaction(testnet, _body as TransferCeloOrCeloErc20Token, (await this.getFirstNodeUrl(chain, testnet)));
+                break;
+            // case Currency.TRON:
+            //     txData = await prepareTronTrc20SignedTransaction(testnet, _body as TransferTronTrc20);
+            //     break;
+            default:
+                throw new Erc20Error(`Unsupported chain ${chain}.`, 'unsuported.chain');
+        }
+        if (body.signatureId) {
+            return {signatureId: await this.storeKMSTransaction(txData, chain, [body.signatureId], body['index'])};
```
Comment (tatumio) [inline erc20-connector/module/src/Erc20Service.ts:145]: I dont like body['index'], its old accessing. use body.index everywhere

**Review by tatumio — CHANGES_REQUESTED** [review]

**Review by tatumio — APPROVED** [review]

**Pull request merged**